### DISCUSSION TITLE
Version 14 Queries

### DISCUSSION THREAD

**Discussion by willpower232** [owen-it/laravel-auditing#980] [General]
@erikn69 @parallels999 (sorry I couldn't find you both on discord so I thought this was the easiest option)

I've spent a bit of time on the v14-dev branch and phpstan can get to level 8 although I haven't tested this in a real project so might need some further tweaks.

I am a little confused about php-cs-fixer, I've never been able to configure it as I expect so normally I use squizlabs/php_codesniffer or easy-coding-standard. The main difference is the [space around the `.` operator](https://www.php-fig.org/psr/psr-12/#62-binary-operators) which php-cs-fixer does not enforce but the other two do so I'd like to switch to something more predictable if that is okay with you.

Most confusingly I don't understand the reasoning behind moving almost all of src/Auditable.php into traits, making it much more difficult to follow the git history and also presenting a few challenges to phpstan as there are a few gaps in the interfaces. Do you know any of the reasoning behind it or should it go back into the one file? The separation also makes merging the master branch next to impossible and it would probably be easier to start again if it is to remain.

It would be good to get this wrapped up well before Laravel 12 so let me know what you think (or kick me on discord where I'm at the same username).

**Comment by parallels999** [comment]
>Do you know any of the reasoning behind it or should it go back into the one file?

We were all added at the same time, the only one who could answer is @MortenDHansen

>if that is okay with you

I don't think it makes much difference, what you feel more comfortable with.

>The separation also makes merging the master branch next to impossible and it would probably be easier to start again if it is to remain.

If there is no feature that is lost, I don't see a problem either, but it seems to me that if something was being added, it would be a matter of adding it again if so

**Comment by angeljqv** [comment]
Lumen support should already be removed in V14

**Reply by willpower232** [reply]
I think the only lumen specific code was removed in #916 so shouldn't be a problem either way although should probably be explicitly stated somewhere.

**Reply by erikn69** [reply]
Only remnants
https://github.com/owen-it/laravel-auditing/blob/2d5f2b1fe49f6e3b07967f240ccd70ef14056ccc/composer.json#L64-L66
https://github.com/search?q=repo%3Aowen-it%2Flaravel-auditing%20lumen&type=code
https://github.com/search?q=repo%3Aowen-it%2Flaravel-auditing.com%20lumen&type=code

**Comment by willpower232** [comment]
I have made a few minor PRs which should be merged but don't need to be shipped

- #985
- #986
- #987
- #988 

Having reverted the rearranging of the code into concerns, it seems the changes in v14 are as follows

- rm support for Laravel < 9 and PHP < 8
  - use newer factory classes
  - ~rm assert class thing fc2acee~ #989 
  - anonymous migration class b90081d
- rm deprecated contracts aaeeafc
- add phpstan/larastan, looks like should be able to achieve level 8
- most excitingly some refactoring around resolving attributes 62df551 and f9cf1bb

cc @erikn69 just so you know

**Reply by willpower232** [reply]
todo double check if https://github.com/owen-it/laravel-auditing/issues/588 is still an issue with the updated code

**Reply by erikn69** [reply]
Hi, great work 

fc2acee: IMO, this is a change for v13, when we went from v6 to v7 it should have been removed, i mean **remove v6 support**(i could be wrong because I haven't checked it.)

**UPDATE:**
https://github.com/laravel/framework/blob/6.x/src/Illuminate/Foundation/Testing/Assert.php
https://github.com/laravel/framework/blob/7.x/src/Illuminate/Testing/Assert.php

**Reply by parallels999** [reply]
>rm support for Laravel < 9 and PHP < 8

I think v14 should remove Laravel < 11.x and PHP < 8.2, Laravel 9.x, 10.x have already reached their EOL by this time(_also there were significant changes in Laravel between 10.x and 11.x_)

![image](https://github.com/user-attachments/assets/094cabcd-b847-45ba-b7f8-2313913344e5)

v13 could be left with active development for a prudent period for Laravel 9.x, 10.x

**Reply by willpower232** [reply]
I understand what you're saying but technically the code of the auditing library doesn't actually change for the newer versions of laravel, I think those laravel versions got dropped because the minimum version of php got changed.

I did consider changing the minimum php version to 8.1, which could then be an argument to drop laravel 9 but it never feels like there is a solid reason to change the minimum version being supported unless the code is dramatically changed too.

If there is some refactoring which makes the code easier to develop on which happens to require a minimum php (or laravel) version change then its much easier to justify officially dropping support for older versions.

Laravel 8 allowed installing on php 7 which means nullsafe operator, constructor property promotion, and all the other cool things that happened could not be used. I haven't double checked the v14 branch but I presume some of those could be added in if they weren't already.

**Reply by parallels999** [reply]
For that you could only remove support in v13 itself, in my opinion removing support for an obsolete version is not a breaking change

**Comment by willpower232** [comment]
I've made some phpstan-related commits. Unfortunately it has become much stricter in the past few weeks since I last worked on this so its a bit harder to figure out all the right answers. I've made PRs for the majority of it and hopefully will have time to have another go soon. I don't think its worth officially adding phpstan just yet until they're all solved.
